modules/bot-improvement/src/backend/db.ts

```diff
@@ -0,0 +1,177 @@
+  initialize: () => Promise<void>
+  getFeedbackItems: (botId: string) => Promise<FeedbackItem[]>
+  updateFeedbackItem: (props: UpdateFeedbackItemProps) => Promise<void>
+  getMessageGroups: (sessionId: string) => Promise<MessageGroup[]>
+}
+
+const isQna = (event: IO.IncomingEvent): boolean => {
+  const intentName = getIntentName(event)
+  return intentName.startsWith(QNA_IDENTIFIER)
+}
+
+const getQnaIdFromIntentName = (intentName: string): string => {
+  return intentName.replace(QNA_IDENTIFIER, '')
+}
+
+const getIntentName = (event: IO.IncomingEvent): string => {
+  return event.nlu.intent.name
+}
+
+const getQnaItemFromEvent = (event: IO.IncomingEvent, qnaItems: QnAItem[]): QnAItem => {
+  const intentName = getIntentName(event)
+  const qnaId = getQnaIdFromIntentName(intentName)
+  return qnaItems.find(item => item.id === qnaId)
+}
+
+const convertStoredEventToMessage = (storedEvent: IO.StoredEvent): Message => {
+  const event = storedEvent.event
+  const payload = event.payload || {}
+  const text = event.preview || payload.text || (payload.wrapped && payload.wrapped.text)
+  const direction = event.direction === 'outgoing' ? 'out' : 'in'
+
+  let source: 'user' | 'bot' = 'user'
+  if (direction === 'out') {
+    source = 'bot'
+  }
+
+  return {
+    id: storedEvent.id,
+    sessionId: storedEvent.sessionId,
+    type: event.type,
+    text,
+    raw_message: event.payload,
+    direction,
+    source,
+    ts: event.createdOn
+  }
+}
+
+export default (bp: SDK): Database => {
+  const knex = bp.database
+
+  const initialize = async () => {
+    if (!knex) {
+      throw new Error('you must initialize the database before')
+    }
+
+    await knex.createTableIfNotExists('bot_improvement_feedback_items', table => {
+      table.increments('id').primary()
+      table
+        .integer('eventId')
+        .unsigned()
+        .notNullable()
+      table.string('status').notNullable()
+      table.string('correctedActionType').notNullable()
+      table.string('correctedObjectId').notNullable()
+    })
+  }
+
+  const getFeedbackItems = async (botId: string) => {
+    const flaggedEvents = await knex
+      .column(
+        { eventId: 'events.id' },
+        'events.event',
+        'events.sessionId',
+        'bot_improvement_feedback_items.status',
+        'bot_improvement_feedback_items.correctedActionType',
+        'bot_improvement_feedback_items.correctedObjectId'
+      )
+      .select()
+      .from('events')
+      .leftJoin('bot_improvement_feedback_items', {
+        'events.id': 'bot_improvement_feedback_items.eventId'
+      })
+      .where({ botId, feedback: -1 })
+      .then(rows =>
+        rows.map(storedEvent => ({
+          ...storedEvent,
+          event: knex.json.get(storedEvent.event)
+        }))
+      )
+
+    const axiosConfig = await bp.http.getAxiosConfigForBot(botId, { localUrl: true })
+    const qnaItems: QnAItem[] = (await axios.get('/mod/qna/questions', axiosConfig)).data.items
```

Comment:
Is there a way to check if a given module (here, QnA) is activated when I start my module? That would be a more elegant solution.

Reply:
Check botpress config

Reply:
You will still not know if the user has access to the module. Maybe QNA is enabled, but the user doesn't have read access to it. That's why I suggest wrapping it in a try/catch. 

There is the AccessControl component in the botpress/utils package which can be used in the frontend. You could also check in the sdk for bp.http.needPermission or bp.http.hasPermission.

Either way, you would still need to check if the module is enabled + if the user has permission. Wrapping in a try catch is easier for now and more safe, since if there's an error with the module, it will not impact everything